lib/membrane/core/element/demand_controller/auto_flow_utils.ex

```diff
@@ -97,14 +129,85 @@ defmodule Membrane.Core.Element.DemandController.AutoFlowUtils do
+  def pop_queues_and_bump_demand(%State{popping_queue?: true} = state), do: state
+
+  def pop_queues_and_bump_demand(%State{} = state) do
+    %{state | popping_queue?: true}
+    |> bump_demand()
+    |> pop_auto_flow_queues_while_needed()
+    |> bump_demand()
+    |> Map.put(:popping_queue?, false)
+  end
+
+  defp bump_demand(state) do
+    if state.effective_flow_control == :pull and
+         MapSet.size(state.satisfied_auto_output_pads) == 0 do
+      state.pads_data
+      |> Enum.flat_map(fn
+        {ref, %{direction: :input, flow_control: :auto, end_of_stream?: false}} -> [ref]
+        _other -> []
+      end)
+      |> auto_adjust_atomic_demand(state)
+    else
+      state
+    end
+  end
+
+  @spec pop_auto_flow_queues_while_needed(State.t()) :: State.t()
+  def pop_auto_flow_queues_while_needed(state) do
+    if (state.effective_flow_control == :push or
+          MapSet.size(state.satisfied_auto_output_pads) == 0) and
+         MapSet.size(state.awaiting_auto_input_pads) > 0 do
+      pop_random_auto_flow_queue(state)
+      |> pop_auto_flow_queues_while_needed()
+    else
+      state
+    end
   end
 
-  defp atomic_demand_positive?(pad_ref, state) do
-    atomic_demand_value =
-      PadModel.get_data!(state, pad_ref, :atomic_demand)
-      |> AtomicDemand.get()
+  defp pop_random_auto_flow_queue(state) do
+    pad_ref = Enum.random(state.awaiting_auto_input_pads)
 
-    atomic_demand_value > 0
+    state
+    |> PadModel.get_data!(pad_ref, :auto_flow_queue)
+    |> Qex.pop()
+    |> case do
+      {{:value, {:buffer, buffer}}, popped_queue} ->
+        state = PadModel.set_data!(state, pad_ref, :auto_flow_queue, popped_queue)
+        state = BufferController.exec_buffer_callback(pad_ref, [buffer], state)
+        pop_stream_formats_and_events(pad_ref, state)
+
+      {:empty, _empty_queue} ->
+        Map.update!(state, :awaiting_auto_input_pads, &MapSet.delete(&1, pad_ref))
+    end
   end
+
+  defp pop_stream_formats_and_events(pad_ref, state) do
+    PadModel.get_data!(state, pad_ref, :auto_flow_queue)
+    |> Qex.pop()
+    |> case do
+      {{:value, {:event, event}}, popped_queue} ->
+        state = PadModel.set_data!(state, pad_ref, :auto_flow_queue, popped_queue)
+        state = EventController.exec_handle_event(pad_ref, event, state)
+        pop_stream_formats_and_events(pad_ref, state)
+
+      {{:value, {:stream_format, stream_format}}, popped_queue} ->
+        state = PadModel.set_data!(state, pad_ref, :auto_flow_queue, popped_queue)
+        state = StreamFormatController.exec_handle_stream_format(pad_ref, stream_format, state)
+        pop_stream_formats_and_events(pad_ref, state)
+
+      {{:value, {:buffer, _buffer}}, _popped_queue} ->
+        state
+
+      {:empty, _empty_queue} ->
+        Map.update!(state, :awaiting_auto_input_pads, &MapSet.delete(&1, pad_ref))
+    end
+  end
+
+  defp output_auto_demand_positive?(%State{satisfied_auto_output_pads: pads}),
+    do: MapSet.size(pads) == 0
```

Comment:
Pattern match is faster than looping over whole map 

```suggestion
  defp output_auto_demand_positive?(%State{satisfied_auto_output_pads: %{}}), do: true
  defp output_auto_demand_positive?(_state), do: false
```